### PR TITLE
op-node,kona: validate span batch overlap content against the safe chain in Holocene

### PR DESCRIPTION
**Pull request by karlfloersch**
## Description

Alternative to #22214 — same defect, simpler rule. This PR now carries the rule in **both clients** (op-node and kona), plus an action proofs test that pins their equivalence with the fault proof program.

### The defect

The Holocene span batch prefix checks (`checkSpanBatchPrefix`) do not validate the contents of batch elements that overlap the safe chain — unlike the legacy full checks (`checkSpanBatch`), which fetch every overlapped payload and compare transactions. Dropping that comparison was sound under a pre-interop invariant: the safe chain was derived from this same batch data, so overlap content provably agrees with it.

Interop deposits-only block replacement breaks that invariant. After a denied block is replaced, the canonical block at the replaced height no longer matches the batch element it was originally derived from — and the prefix checks never notice:

- `GetSingularBatches` past-skips every element at or below the safe head without inspecting content, including the since-replaced (denied) block.
- The deny check (`IsDenied`) only fires when a payload at the denied height is *proposed* (`engine/payload_process.go`), and the channel flush is a side effect of that deny event (`DepositsOnlyAttributes`). A walk anchored **at** the replacement never proposes the denied block, so neither fires.
- The span's tail is then re-stamped onto the replacement and applied — splicing the remainder of an invalidated lineage onto the canonical chain.

The result is that the derived chain depends on where the pipeline (re-)anchored relative to the replacement: an anchor **below** the denied block re-proposes it → deny → replacement + channel flush → the next channel wins; an anchor **at** the replacement past-skips it → the same channel's tail wins. Same L1 data, same deny list, different chains. This was observed on devnet `interop-reorg-5`, where a post-invalidation reset anchored a live supernode at the replacement and flipped it onto a lineage that from-genesis syncs can never derive (writeup: internal doc "Double Batch Derivation Issue"). PR #21970 closed the analogous hole on the consolidation path, but both existing deny hooks are keyed on a payload/attributes existing at the denied height — on this path, nothing at the denied height is ever processed.

kona has the identical gap: its Holocene `BatchStream` runs only `check_batch_prefix`, while the full overlap comparison sits unused in the legacy `check_batch` path.

### The fix

Reinstate the **pre-Holocene overlap rule in full**, in both clients: a span batch that overlaps the safe chain must agree with the safe chain, at *every* overlapped height — transactions and L1 origin number, the exact legacy comparison. A mismatch drops the batch and flushes the channel; a payload fetch failure is Undecided.

Structure (settled after two rounds of review — the interim fold of the overlap check into a combined Holocene check function was reverted because it moved overlap payload reads ahead of the deterministic legacy checks pre-Holocene):

- **op-node**: the legacy loop is factored verbatim out of `checkSpanBatch` into `checkSpanBatchOverlap`. `checkSpanBatchPrefix` is untouched. The legacy `checkSpanBatch` calls the overlap check at its original tail position (preserving pre-Holocene check order and error/liveness behavior exactly), and `BatchStage` runs it as a second phase after an accepted prefix check, guarded on the batch actually overlapping the safe chain.
- **kona**: the mirror-image structure — `check_batch_overlap` factored out of `check_batch` and called at its legacy tail; `check_batch_prefix` untouched; `BatchStream` runs the overlap check as a second phase after an accepted prefix check. The check-batch-prefix duration metric covers only the prefix checks, as before.

This directly enforces the invariant the prefix-only checks silently dropped, and restores anchor-invariance of derivation across replacement boundaries: re-deriving from any anchor yields the same chain, because a batch carrying a since-replaced block is rejected on every walk instead of being past-skipped into a splice.

### Fault proof cost (review round 2)

The overlap check's per-block `block_by_number` lookups resolved by walking headers back from the L2 safe head in the fault-proof provider, making an overlap of length N cost O(N²) header preimage reads. `OracleL2ChainProvider` now carries the same by-number cache its L1 counterpart already had: walks record every header they discover, lookups resume from the closest cached ancestor, and the ascending overlap sweep costs one walk over the range in total. Provider-level regression tests pin the complexity with a counting mock oracle (the five-deep ascending sweep is asserted at exactly 5 header reads; the uncached walk did 15).

## Commit structure (red → green)

The first commit adds only the action proofs test; each subsequent commit turns part of it green, so the defect and the fix can be replayed per client:

1. **`kona: add span batch overlap content action test`** — test only. Red in both clients: op-node splices the invalidated lineage (safe-head assertions fail); kona-client would reject the honest claim at the divergent height.
2. **`op-node: Validate span batch overlap content against the safe chain in Holocene`** — Go fix + Go unit tests.
3. **`op-node: fold overlap check into checkSpanBatchHolocene (review feedback)`** — interim restructure, reverted in commit 8.
4. **`kona: validate span batch overlap content against the safe chain post-Holocene`** — kona fix + kona unit test updates. Fully green.
5. **`kona: address action test review comments`** — scenario rework (batcher equivocation with alternative valid blocks; fork-variable across Fjord/Holocene/latest), `AddDefaultTestCases`.
6. **`op-node,kona: restore displaced test coverage, tighten comments (self-review)`**, **`op-node,kona: correct stale retry-later comments on Undecided batches`** — self-review fixes.
7. *(review round 2)* **`op-node: revert overlap fold, restore prefix + stage-level overlap checks`** — reverts commit 3: the fold moved overlap payload reads ahead of the legacy full checks pre-Holocene, so a deterministically invalid batch could sit Undecided at the queue front during an engine read outage. The restored functions are character-identical to commit 2 (modulo the self-review comment corrections), i.e. the shape the original testing validated.
8. **`kona: mirror op-node revert — restore check_batch_prefix, stage-level overlap check`** — the Rust mirror, keeping the clients in lockstep.
9. **`kona-proof: linear by-number header walks for span batch overlap validation`** — the fault-proof complexity blocker: by-number header cache + counting-oracle regression tests + the encode-buffer hoist nit.
10. **`op-node,kona: strengthen overlap flush tests with unread sentinels and a multi-block case`** — flush tests now prove the channel was actually flushed (unread sentinel discarded), and a two-block-overlap case exercises the comparison loop beyond its first iteration.
11. **`kona: action test expects the stage-level overlap drop message`** — the action test's sequencer log expectation follows the restored drop message.

## The action proofs test

`rust/kona/tests/proofs/span_batch_overlap_test.go` (`TestSpanBatchOverlapContent`) pins the rule end-to-end via batcher equivocation with alternative *valid* blocks (per review: independent of deposits-only replacement rules, so it runs across forks — Fjord, Holocene, and latest):

1. Channel 1 carries span `[1, 2]`: the canonical chain, derived to the safe head.
2. Channel 2 carries span `[2*, 3*]` — the same heights with user transactions stripped, an alternative valid lineage. Its overlap disagrees with the safe chain at height 2: it must be dropped whole (the safe head stays at 2) instead of past-skipping `2*` and splicing `3*` onto canonical 2.
3. Channel 3 carries the canonical block 3: it must derive, instead of being dropped as a past batch because height 3 is already wrongly occupied by `3*`.
4. `RunFaultProofProgramFromGenesis` validates honest claims from the op-node-derived chain at every height 0..3, plus junk-claim variants. Height 3 is where an unfixed client derives `3*` instead of the canonical block and rejects the honest claim.

Verified locally at each commit boundary: at the test-only commit, **Holocene and latest-fork variants are red** (op-node splices the equivocating lineage) while **Fjord is green** — pinning that the pre-Holocene overlap rule holds unchanged on the BatchQueue path and the defect is precisely the Holocene gap. At the Go-fixed/kona-unfixed boundary, kona-client rejects the honest claim at height 3 (`ErrClaimNotValid`). At head, all six fork/claim subtests are green (re-verified after the round-2 restructure).

## Trade-offs vs #22214

**Simpler:**
- The rule is a pure function of the batch and the local chain — no `SuperAuthority` interface change, no deny list plumbing, no supernode changes, no external state to define. Now implemented in both clients symmetrically.

**Costs:**
- **Consensus-visible on every Holocene chain**, not just interop chains: a batch whose overlap disagrees with the safe chain at *any* height (producible today only by batcher equivocation) is now dropped + flushed instead of having its tail applied. This changes which continuation wins in equivocation scenarios, so it needs a spec decision before it ships anywhere real (spec draft: ethereum-optimism/specs#928). op-node/kona parity is in-PR (both clients change in lockstep); the fault proof program picks the change up via kona-client.
- **O(overlap) engine queries per overlapping span batch**: every pipeline reset walk-back pays `PayloadByNumber` for each overlapped block of the span straddling the safe head (typically one span per walk; roughly the pre-Holocene restart cost). Fully-past spans still short-circuit via `BatchPast` before the check. On the fault-proof path this is now linear in the overlap length (see above).

## Test changes

- `TestSpanBatchOverlapContent` (new action proofs test) — see above.
- `TestBatchStage_OverlapContent` (new, Go unit) pins the incident shape at the stage level: a span carrying a replaced block is dropped with a channel flush **that provably discards the rest of the channel** (unread sentinel), a span matching the canonical chain is accepted, payload fetch errors are Undecided, and a two-block overlap diverging only at its second element is caught by the comparison loop's later iterations.
- kona `test_overlap_mismatch_drops_span_and_flushes_unread_batches` mirrors the sentinel + multi-block case at the `BatchStream` level (the test provider's `flush` now actually discards queued batches instead of being a silent no-op).
- `OracleL2ChainProvider` regression tests: ascending five-deep overlap sweep costs exactly 5 header preimage reads (was 15), resume-from-cached-ancestor, repeat-hit, and past-head cases.
- `go test ./op-node/rollup/...`, `cargo nextest run -p kona-protocol -p kona-derive -p kona-proof`, clippy, rustfmt, and `check-no-std` pass.

## Which one to land

Proposal: use #22214 as the immediately-deployable fix for the interop devnets (no consensus impact outside the deny machinery's footprint), and treat this PR as the concrete artifact for the protocol-level discussion — if the spec adopts "overlap must agree with the safe chain" (ethereum-optimism/specs#928), this is the implementation in both clients, with the equivalence test to hold them together.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Review note: the moved loops are verbatim

In both clients the overlap loop is lifted character-for-character (dedented) out of the legacy full-check function into the new shared function, with zero logic changes — verified by diffing the regions with leading whitespace normalized. The only post-move deviation is the review-requested hoist of the transaction encoding buffer in kona's `check_batch_overlap`. To see it directly:

```
git diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space develop -- op-node/rollup/derive/batches.go rust/kona/crates/protocol/protocol/src/batch/span.rs
```

### CODE REVIEW

File: op-node/rollup/derive/batch_stage.go [hunk 1]
```diff
@@ -155,6 +155,25 @@ func (bs *BatchStage) nextSingularBatchCandidate(ctx context.Context, parent eth
 			return nil, NewCriticalError(errors.New("impossible future batch validity"))
 		}
 
+		// The prefix checks do not validate overlap contents against the safe chain. A span batch
+		// that disagrees with the safe chain — possible since interop block replacement — must be
+		// dropped as a whole, so that the remainder of an invalidated lineage cannot be spliced
+		// onto the canonical chain. See checkSpanBatchOverlap for details.
+		if parentBlock.Number < parent.Number {
+			switch validity := checkSpanBatchOverlap(ctx, bs.config, bs.Log(), spanBatch, parentBlock, parent, bs.l2); validity {
```
Comment (sebastianst) [inline op-node/rollup/derive/batch_stage.go]: I think we should just move the `checkSpanBatchOverlap` check to the end of `checkSpanBatchPrefix`. Arguably `checkSpanBatchPrefix` isn't then a great name any more, but it's the full set of span batch checks that need to run post-Holocene. Maybe we then rename it to `checkSpanBatchHolocene`.

Comment (karlfloersch) [reply op-node/rollup/derive/batch_stage.go]: @sebastianst I think that I faithfully implemented your recommendation. Does this look about right? Definitely seems to be a simplification

File: rust/kona/tests/proofs/span_batch_overlap_test.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		helpers.ExpectNoError(),
+	)
+	matrix.AddTestCase(
+		"JunkClaim",
+		testCase{},
+		helpers.NewForkMatrix(helpers.Holocene, helpers.LatestFork),
+		runTest,
+		helpers.ExpectError(helpers.ErrClaimNotValid),
+		helpers.WithL2Claim(common.HexToHash("0xdeadbeef")),
+	)
```
Comment (sebastianst) [inline rust/kona/tests/proofs/span_batch_overlap_test.go]: there's a `AddDefaultTestCases...` function

File: rust/kona/tests/proofs/span_batch_overlap_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+//  2. Channel 2 carries the same invalidated lineage [1, 2*, 3] again — the incident's
+//     "re-derivation walk" over a batch containing a since-replaced block. Its overlap
+//     (heights 1-2) must be validated against the safe chain: element 2* conflicts with 2'.
+//     With the overlap content rule, the whole batch is dropped and the channel flushed.
+//     Without it, element 3 — built on the invalidated 2* — is spliced onto 2', producing
+//     another deposits-only replacement 3' derived from a lineage that was never canonical.
+//  3. Channel 3 carries the canonical continuation 3” (built on 2', with a live transaction).
+//     With the rule, it applies and the safe head becomes 3”. Without it, height 3 is already
+//     (wrongly) occupied by 3', and the canonical continuation is dropped as a past batch —
+//     the equivocating channel's tail wins over the canonical chain.
```
Comment (sebastianst) [inline rust/kona/tests/proofs/span_batch_overlap_test.go]: I can imagine a simpler scenario
- channel 1 has span `[1, 2]`
- channel 2 has span `[2', 3]` where 2' is an alternative valid block to 2.

without overlap checks, channel 2 will yield 3. with overlap checks, it will yield nothing and safe head stays at 2.

This way the test is independent of invalid payload replacement rules and could even be a variable test that runs over multiple forks, e.g. Fjord, Holocene and _latest_ fork.

File: rust/kona/tests/proofs/span_batch_overlap_test.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		txs := block.Transactions()
+		newTx, err := txs[1].WithSignature(alice, make([]byte, 65))
+		if err != nil {
+			panic(err)
+		}
+		txs[1] = newTx
+		return block
+	}
+
+	runTest := func(gt *testing.T, testCfg *helpers.TestCfg[testCase]) {
```
Comment (sebastianst) [inline rust/kona/tests/proofs/span_batch_overlap_test.go]: I think you can just use `nil` instead of an empty `testCase` struct, check other tests what they do.

But see my other comment, we may want to use the `forks.Name` type here to make this a variable test that we run for three different forks.

File: rust/kona/tests/proofs/span_batch_overlap_test.go (hunk 3 above)
Comment (karlfloersch) [reply rust/kona/tests/proofs/span_batch_overlap_test.go]: Test should now use that methodology!

File: rust/kona/tests/proofs/span_batch_overlap_test.go (hunk 2 above)
Comment (karlfloersch) [reply rust/kona/tests/proofs/span_batch_overlap_test.go]: Done!

File: rust/kona/tests/proofs/span_batch_overlap_test.go (hunk 4 above)
Comment (karlfloersch) [reply rust/kona/tests/proofs/span_batch_overlap_test.go]: Done!

**Comment by karlfloersch** [comment]
**Note for reviewers — `BatchUndecided` on the Holocene stage skips the batch (pre-existing edge case, exposure widened here)**

The overlap content check returns `BatchUndecided` when a canonical payload cannot be fetched right now. On the pre-Holocene `BatchQueue`, Undecided genuinely means "retry later": the batch stays buffered and is re-checked. On the Holocene `BatchStage`/`BatchStream`, the span batch has already been consumed from the channel reader by the time validity comes back, and nothing retains it — an Undecided batch is **skipped, not retried**. A transient engine fetch failure can therefore make a node skip a span batch its peers accept; the remainder of that channel then drops on parent-hash checks, the window eventually force-fills, and the node's safe chain diverges with no self-heal.

This lifecycle predates this PR: the prefix checks' parent-block fetch (`L2BlockRefByNumber`) has had the same shape since Holocene, on exactly the same input class (overlapping span batches). What this PR changes is exposure — the overlap check adds one `PayloadByNumber` per overlapped block, so more fetches can hit a transient failure. The old "retry later" comments at these sites described only the BatchQueue behavior; 048341cedc corrects them to say what actually happens. (The fault proof program is unaffected: its oracle is deterministic, so there is no transient-failure case.)

Fixing it properly means retaining the in-flight span batch until validity resolves. That is a real stage-lifecycle change with a trap: `Undecided` conflates **transient local fetch failures** (retry in place is correct) with **insufficient L1 context** ("need more l1 blocks…"), where retrying in place livelocks — the origin must advance first, and signaling `io.EOF` instead would trigger empty-batch derivation and force-fill past the undecided batch. The BatchQueue's buffer handles both flavors; the Holocene stages need an equivalent pending-slot design, mirrored in kona, with fault-injection tests. Proposing to track that as its own issue/PR rather than growing this one.

File: rust/kona/tests/proofs/span_batch_overlap_test.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+//     an alternative valid lineage. Its overlap (height 2) disagrees with the safe chain, so
+//     the whole batch must be dropped and the safe head stay at 2. Without the overlap content
+//     rule, 2* is past-skipped and 3* — a valid empty block — is spliced onto canonical 2.
+//  3. Channel 3 carries the canonical block 3: it must derive (fixed), instead of being dropped
+//     as a past batch because height 3 is already wrongly occupied by 3* (unfixed) — the
+//     equivocating channel's tail must not win over the canonical chain.
+//
+// The fault proof program then replays the same L1 data with honest claims taken from the
+// op-node-derived chain at every height 0..3. Height 3 is where a client without the overlap
+// rule derives 3* instead of the canonical block 3 and rejects the honest claim, pinning that
```
Comment (sebastianst) [inline rust/kona/tests/proofs/span_batch_overlap_test.go:37]: I can imagine a simpler scenario
- channel 1 has span `[1, 2]`
- channel 2 has span `[2', 3]` where 2' is an alternative valid block to 2.

without overlap checks, channel 2 will yield 3. with overlap checks, it will yield nothing and safe head stays at 2.

File: rust/kona/crates/protocol/protocol/src/batch/span.rs [hunk 6]
```diff
@@ -519,78 +519,95 @@ impl SpanBatch {
...
-                    warn!(
-                        target: "batch_span",
-                        "overlapped block's tx count does not match, safe_block_txs: {}, batch_txs: {}",
-                        safe_block.transactions.len(),
-                        batch_txs.len()
-                    );
-                    return BatchValidity::Drop(BatchDropReason::OverlappedTxCountMismatch);
+        for i in 0..(l2_safe_head.block_info.number - parent_num) {
+            let safe_block_num = parent_num + i + 1;
+            let safe_block_payload = match fetcher.block_by_number(safe_block_num).await {
```
Comment (sebastianst) [inline rust/kona/crates/protocol/protocol/src/batch/span.rs:555]: Blocking: in the fault-proof provider, each `block_by_number` call resolves `header_by_number` by walking backward from the unchanged L2 safe head. Checking an overlap of length `N` therefore performs `N + (N-1) + ... + 1` header preimage reads/decodes, plus `N` transaction-trie reconstructions. A temporary counting-oracle test produced 15 header reads for `N=5`.

Please make the proof path linear, e.g. with a range/backward-walk API or by caching/indexing headers discovered during the first walk, and pin the complexity with a provider-level regression test. op-node's direct payload-by-number lookups do not have this asymmetry.

File: rust/kona/crates/protocol/protocol/src/batch/span.rs [hunk 7]
```diff
@@ -519,78 +519,95 @@ impl SpanBatch {
...
+                    "overlapped block's tx count does not match, safe_block_txs: {}, batch_txs: {}",
+                    safe_block.transactions.len(),
+                    batch_txs.len()
+                );
+                return BatchValidity::Drop(BatchDropReason::OverlappedTxCountMismatch);
+            }
+            let batch_txs_len = batch_txs.len();
+            #[allow(clippy::needless_range_loop)]
+            for j in 0..batch_txs_len {
+                let mut buf = Vec::new();
```
Comment (sebastianst) [inline rust/kona/crates/protocol/protocol/src/batch/span.rs]: Small local optimization: hoist this `Vec` outside the inner loop and `clear`/reuse it for each transaction. That removes one allocation per compared transaction.

We considered comparing transaction hashes instead. The safe `OpTxEnvelope` already carries its hash, but the span side would still require a Keccak over every candidate transaction. Direct byte comparison is cheapest when both raw encodings are available, but `BatchValidationProvider` currently returns a decoded `OpBlock`, and the RPC provider does not retain raw transaction bytes. Buffer reuse improves the current representation without adding a provider API or extra hashing.

File: op-node/rollup/derive/batches.go [hunk 8]
```diff
@@ -324,7 +333,7 @@ func checkSpanBatchPrefix(ctx context.Context, cfg *rollup.Config, log log.Logge
 func checkSpanBatch(ctx context.Context, cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l2SafeHead eth.L2BlockRef,
 	batch *SpanBatch, l1InclusionBlock eth.L1BlockRef, l2Fetcher SafeBlockFetcher,
 ) BatchValidity {
-	prefixValidity, parentBlock := checkSpanBatchPrefix(ctx, cfg, log, l1Blocks, l2SafeHead, batch, l1InclusionBlock, l2Fetcher)
+	prefixValidity, parentBlock := checkSpanBatchHolocene(ctx, cfg, log, l1Blocks, l2SafeHead, batch, l1InclusionBlock, l2Fetcher)
```
Comment (sebastianst) [inline op-node/rollup/derive/batches.go]: This also moves the overlap reads ahead of the remaining legacy full checks on pre-Holocene batches. A batch with a deterministically invalid new element used to be dropped before overlap payload reads; now a persistent local read failure returns `BatchUndecided` first and keeps the invalid batch at the queue front indefinitely.

Please keep overlap validation at the legacy full-check tail while having the Holocene path call prefix checks followed by overlap checks. That preserves deterministic acceptance equivalence without changing pre-Holocene error/liveness behavior.

**Review by sebastianst — CHANGES_REQUESTED** [review]
The successful-read accept/drop behavior is aligned between op-node and kona. I found one fault-proof complexity blocker, plus two liveness/test issues worth fixing here.

The Holocene wording does imply that overlap checks stopped. I opened ethereum-optimism/specs#928 to make retaining the Delta overlap checks explicit and to remove wording that suggests otherwise. This is a retroactive protocol change; that draft records the current safety assumption and the need for a historical production-data scan.

The known `BatchUndecided` lifecycle bug is pre-existing, but this change widens its exposure from one parent lookup to up to one payload lookup per overlapped block. A failed read consumes the in-flight span in both clients, so the post-safe tail is lost until reset/replay. I agree this does not change fault-proof outputs when authenticated oracle data is available, but it does increase local derivation-split risk. However, we accept this tradeoff since we want to improve `BatchUndecided` handling in a follow up and for kona, the IO-free derivation refactor will remove this defect by design.

Please also strengthen the channel-flush test: add an unread sentinel batch to the same channel (or instrument `FlushChannel`) and assert it is discarded. The current log/final-head assertions can pass if the code logs a flush without actually flushing. A multi-block overlap case where an early block matches and a later transaction differs would cover the loop beyond `N=1`.

Verification: `go test ./op-node/rollup/derive -count=1`; `cargo test -p kona-protocol -p kona-derive -j 6`; temporary counting-oracle experiment confirming 15 header preimage reads for five sequential overlap lookups.

🤖 *Co-created with openai-codex/gpt-5.6-sol*

File: op-node/rollup/derive/batches.go (hunk 8 above)
Comment (sebastianst) [reply op-node/rollup/derive/batches.go]: I think a simple way to achieve this is to restore the `checkSpanBatchPrefix` function as it was before, then simply have `checkSpanBatchHolocene` call `checkSpanBatchPrefix` followed by `checkSpanBatchOverlap`. Then the pre-Holocene `checkSpanBatch` function can call (like before) `checkSpanBatchPrefix`, then perform its full checks, then at the end call `checkSpanBatchOverlap`.

**Review by sebastianst — APPROVED** [review]
